node/pkg/telemetry/loki.go

```diff
@@ -142,15 +163,133 @@ func NewLokiCloudLogger(ctx context.Context, logger *zap.Logger, url string, pro
+	go logWriter(workerContext, localLogger, bufferedChan, workerExitedC, c)
+
 	return &Telemetry{
 		encoder: &guardianTelemetryEncoder{
 			Encoder: zapcore.NewJSONEncoder(zapdriver.NewProductionEncoderConfig()),
 			logger: &ExternalLoggerLoki{
-				c:           c,
-				labels:      lokiLabels,
-				localLogger: localLogger,
+				c:             c,
+				labels:        lokiLabels,
+				localLogger:   localLogger,
+				bufferedChan:  bufferedChan,
+				cancelWorker:  cancelWorker,
+				workerExitedC: workerExitedC,
 			},
 			skipPrivateLogs: skipPrivateLogs,
 		},
 	}, nil
 }
+
+// logWriter is the go routine that takes log messages off the buffered channel and posts them to the Loki client. It can block
+// on the Loki client until our context is canceled, meaning we are shutting down. On shutdown, it tries to flush buffered messages
+// and shutdown the Loki client using a timeout for both actions.
+func logWriter(ctx context.Context, logger *zap.Logger, localC chan api.Entry, workerExitedC chan struct{}, c client.Client) {
+	// pendingEntry is used to save the last log message if the write to Loki is interrupted by the context being canceled. We will attempt to flush it.
+	var pendingEntry *api.Entry
+
+	for {
+		select {
+		case entry, ok := <-localC:
+			if !ok {
+				logger.Error("Loki log writer is exiting because the buffered channel has been closed")
+				return
+			}
+
+			// Write to Loki in a blocking manner unless we are signaled to shutdown.
+			select {
+			case c.Chan() <- entry:
+				pendingEntry = nil
+			case <-ctx.Done():
+				// Time to shutdown. We probably failed to write this message, save it so we can try to flush it.
+				pendingEntry = &entry
+			}
+		case <-ctx.Done():
+			logger.Info("Loki log writer shutting down")
+
+			// Flush as much as we can in our allowed time.
+			if err := flushLogsWithTimeout(localC, c, pendingEntry); err != nil {
+				logger.Error("worker failed to flush logs", zap.Error(err), zap.Int("numEventsRemaining", len(localC)))
+			}
+
+			// Stop the client without blocking indefinitely.
+			if err := stopClientWithTimeout(c); err != nil {
+				logger.Error("worker failed to stop Loki client", zap.Error(err))
+			}
+
+			// Signal that we are done.
+			workerExitedC <- struct{}{}
+			logger.Info("Loki log writer exiting")
+			return
+		}
+	}
+}
+
+// flushLogsWithTimeout is used to flush any buffered log messages on shutdown.
+// It uses a timeout so that we only delay guardian shutdown for so long.
+func flushLogsWithTimeout(localC chan api.Entry, c client.Client, pendingEntry *api.Entry) error {
+	// Create a timeout context. Base it on the background one since ours has been canceled.
+	timeout, cancel := context.WithTimeout(context.Background(), clientTimeout)
+	defer cancel()
+
+	if pendingEntry != nil {
+		select {
+		case c.Chan() <- *pendingEntry:
+		case <-timeout.Done():
+			return errors.New("timeout")
```

Comment:
I think it's worth disambiguating the error messages here. e.g. this could be "timeout: could not write pending entry" and the other one "timeout: could not completely flush buffered logs".

It might be helpful also write out the configured timeout so that it's visible in the logs as something like "timeout after X ms"

Reply:
I make the text unique. I don't think it's necessary to log the timeout since it is a constant in the code.